### ISSUE TITLE
Error: Cuda("no cuda implementation for softmax-last-dim")

### ISSUE THREAD

**Issue by hezhaozhao-git** [huggingface/candle#1330]
- carogo build error:
   ` Error: Cuda("no cuda implementation for softmax-last-dim")`
- nvidia-smi:
 
![image](https://github.com/huggingface/candle/assets/46663318/992cc126-80b7-4fa8-a08c-8b7d17a45d4d)

- nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Sep__8_19:17:24_PDT_2023
Cuda compilation tools, release 12.3, V12.3.52
Build cuda_12.3.r12.3/compiler.33281558_0

**Comment by LaurentMazare** [comment]
Yes the backward pass is not supported for this operator but it is for the standard softmax (which is a bit slower though), so you can probably just do `x.softmax(D::Minus1)` to get around this error.

**Comment by danielclough** [comment]
I am getting this as a runtime error from `candle_transformers::models::mistral::Model`.

## Errors (abridged):
```
...
   2: mistral::TextGeneration::run
             at ./src/main.rs:89:38
   3: mistral::main
             at ./src/main.rs:261:5
...
```

```rs
// ./src/main.rs:89:38
88: let logits = match &mut self.model {
89:    Model::Mistral(m) => m.forward(&input, start_pos)?,
90:    Model::Quantized(m) => m.forward(&input, start_pos)?,
91: };
```

```rs
// ./src/main.rs:261:5
251: let mut pipeline = TextGeneration::new(
252:        model,
253:        tokenizer,
254:        args.seed,
255:        args.temperature,
256:        args.top_p,
257:        args.repeat_penalty,
258:        args.repeat_last_n,
259:        &device
260:    );
261:    pipeline.run(&args.prompt, args.sample_len)?;
```

## Source:

I guess the problem comes from here:
https://github.com/huggingface/candle/blob/f4fcf6090045ac44122fd5f0a7e46db6e3e16528/candle-transformers/src/models/mistral.rs#L272C50-L272C50

```rs
// candle/candle-transformers/src/models/mistral.rs
272:            let attn_weights = candle_nn::ops::softmax_last_dim(&attn_weights)?;
```

**Comment by hezhaozhao-git** [comment]
> I ~am~ was getting this as a runtime error from `candle_transformers::models::mistral::Model`.
> 
> ## Edit (solution):
> I solved this problem by adding `features = ["cuda"]` to `candle-transformers`.
> 
> ## Errors (abridged):
> ```
> ...
>    2: mistral::TextGeneration::run
>              at ./src/main.rs:89:38
>    3: mistral::main
>              at ./src/main.rs:261:5
> ...
> ```
> 
> ```rust
> // ./src/main.rs:89:38
> 88: let logits = match &mut self.model {
> 89:    Model::Mistral(m) => m.forward(&input, start_pos)?,
> 90:    Model::Quantized(m) => m.forward(&input, start_pos)?,
> 91: };
> ```
> 
> ```rust
> // ./src/main.rs:261:5
> 251: let mut pipeline = TextGeneration::new(
> 252:        model,
> 253:        tokenizer,
> 254:        args.seed,
> 255:        args.temperature,
> 256:        args.top_p,
> 257:        args.repeat_penalty,
> 258:        args.repeat_last_n,
> 259:        &device
> 260:    );
> 261:    pipeline.run(&args.prompt, args.sample_len)?;
> ```

This method solved the problem, but I didn't use candle-transformers, my code:
`use candle_ext::{candle::{D, DType, Device, Result, Tensor}, TensorExt, F, candle};

// use candle_flash_attn::flash_attn_varlen;

fn main() -> Result<()> {
    let device = Device::new_cuda(0)?;

    let q = Tensor::randn(0., 1., (3, 3, 2, 4), &device)?;
    let k = Tensor::randn(0., 1., (3, 3, 2, 4), &device)?;
    let v = Tensor::randn(0., 1., (3, 3, 2, 4), &device)?;
    let m = Tensor::ones((q.dim(D::Minus2)?, k.dim(D::Minus2)?), DType::U8, &device)?.tril(0)?;
    
    let o = F::scaled_dot_product_attention(&q, &k, &v, Some(&m), None, None, Some(1.0))?;
    println!("{:?}", o);

    // let dims = q.dims().to_vec();
    // let (batch_size, seq_len, n_heads, d) = (dims[0], dims[1], dims[2], dims[3]);
    // let seqlens_q = Tensor::arange_step(0f32, (batch_size as f32 + 1.0) * seq_len as f32, seq_len as f32, &device)?;
    // let seqlens_k = Tensor::arange_step(0f32, (batch_size as f32 + 1.0) * seq_len as f32, seq_len as f32, &device)?;

    // seqlens_q: &Tensor,
    // seqlens_k: &Tensor,
    // max_seqlen_q: usize,
    // max_seqlen_k: usize,
    // softmax_scale: f32,
    // causal: bool,
    // let o1 = flash_attn_varlen(&q, &k, &v, &seqlens_q, &seqlens_k, seq_len , seq_len, 1.0, true);
    // println!("{:?}", o1);
    Ok(())
}`

**Comment by LaurentMazare** [comment]
Closing stale issues, feel free to open a fresh one if you're still having problems.
